### PR TITLE
PCAL64XXA retention

### PR DESCRIPTION
**Pull request by benediktibk**
This PR is a follow-up to #71336.

The main difference is that there is no dedicated driver state retention module anymore, everything is handled with one retention area. The is_valid function is actually sufficient for the PCAL64XXA, as all the current states can be read out from the chip itself.

### CODE REVIEW

**Comment by benediktibk** [comment]
@henrikbrixandersen Would you consider this approach more suitable for upstream than #72688?

File: drivers/gpio/Kconfig.pcal64xxa [hunk 1]
```diff
@@ -18,3 +18,12 @@ config GPIO_PCAL64XXA_INIT_PRIORITY
 	depends on GPIO_PCAL64XXA
 	help
 	  Device driver initialization priority.
+
+config GPIO_PCAL64XXA_RETENTION
+	bool "Device state retention"
+	default n
```
Comment (bjarki-trackunit) [inline drivers/gpio/Kconfig.pcal64xxa:24]: default n is implied

File: drivers/gpio/Kconfig.pcal64xxa [hunk 2]
```diff
@@ -18,3 +18,12 @@ config GPIO_PCAL64XXA_INIT_PRIORITY
 	depends on GPIO_PCAL64XXA
 	help
 	  Device driver initialization priority.
+
+config GPIO_PCAL64XXA_RETENTION
```
Comment (bjarki-trackunit) [inline drivers/gpio/Kconfig.pcal64xxa:22]: I would let the driver handle this automatically using the devicetree, specifically the macro `DT_ANY_INST_HAS_PROP_STATUS_OKAY` which can be used to check if any node of the gpio controller has optional the property `retention` :)

File: drivers/gpio/gpio_pcal64xxa.c [hunk 3]
```diff
@@ -119,6 +125,9 @@ struct pcal64xxa_drv_cfg {
 	const struct gpio_dt_spec gpio_reset;
 	const struct gpio_dt_spec gpio_interrupt;
 	const struct pcal64xxa_chip_api *chip_api;
+#ifdef CONFIG_GPIO_PCAL64XXA_RETENTION
```
Comment (bjarki-trackunit) [inline drivers/gpio/gpio_pcal64xxa.c:128]: See `DT_ANY_INST_HAS_PROP_STATUS_OKAY` to make inclusion automatic

File: drivers/gpio/Kconfig.pcal64xxa (hunk 2 above)
Comment (bjarki-trackunit) [reply drivers/gpio/Kconfig.pcal64xxa:22]:
```
#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(retention) && defined(CONFIG_RETENTION)

#endif
```

**Review by bjarki-andreasen — COMMENTED** [review]
The test suite should be moved to drivers/gpio, as it is testing the gpio controller, not the retention subsystem itself :) Otherwise, every device driver using gpios would be in `tests/drivers/gpio` :) wild that there is no dedicated test suite for retention though.

File: drivers/gpio/Kconfig.pcal64xxa (hunk 1 above)
Comment (benediktibk) [reply drivers/gpio/Kconfig.pcal64xxa]: done (actually removed)

File: drivers/gpio/Kconfig.pcal64xxa (hunk 2 above)
Comment (benediktibk) [reply drivers/gpio/Kconfig.pcal64xxa]: I like the idea, done.

File: drivers/gpio/gpio_pcal64xxa.c (hunk 3 above)
Comment (benediktibk) [reply drivers/gpio/gpio_pcal64xxa.c]: done

**Review by bjarki-andreasen — DISMISSED** [review]
I think this looks good :)

**Comment by benediktibk** [comment]
@henrikbrixandersen Could you please take a look?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
This does not match what we discussed and agreed on in the Architecture WG meeting where this was discussed (expose a reset function for this particular driver, handle retention in your downstream app).

**Comment by benediktibk** [comment]
> This does not match what we discussed and agreed on in the Architecture WG meeting where this was discussed (expose a reset function for this particular driver, handle retention in your downstream app).

Yes, I realized after the discussion that there is another possibility which does not require the exposed reset function. I thought I might propose this approach, as I think it is rather well contained within the driver itself.

**Comment by benediktibk** [comment]
Rebased onto main to resolve merge conflicts.

**Comment by benediktibk** [comment]
@henrikbrixandersen Do you disagree because this approach is different than what was discussed in the architecture WG or do you disagree because of the actual implementation? Once again, I just proposed this version as I didn't have the idea before the meeting and thought it might be a nice option, as it is rather well contained within the driver itself.
If this is also not acceptable for upstream I will go down the road of an externally available reset function.

**Comment by dleach02** [comment]
@henrikbrixandersen do you want this PR brought back to arch wg or can you discuss directly in the PR?

**Comment by benediktibk** [comment]
Rebased onto main to resolve merge conflicts.

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen Do you disagree because this approach is different than what was discussed in the architecture WG or do you disagree because of the actual implementation? Once again, I just proposed this version as I didn't have the idea before the meeting and thought it might be a nice option, as it is rather well contained within the driver itself. If this is also not acceptable for upstream I will go down the road of an externally available reset function.

The consensus in the Architecture WG was not to use the retention subsystem from with the GPIO driver subsystem, but that a driver-specific reset routine (as I recall it?) could be exposed through a driver-specific header file.

This PR does exactly what the Arch WG rejected. If you wish to bring it forward again, please do so, but personally my mind has not changed about this being a layering violation for a single, very specific application and out-of-tree board use-case.

**Pull request closed without merging**